src/utils/constants.hpp

```diff
@@ -0,0 +1,228 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_CONSTANTS_HPP_
+#define UTILS_CONSTANTS_HPP_
+
+#include <cmath>
+#include <string>
+
+namespace parthenon {
+namespace constants {
+
+// Conversion factors from the SI unit system
+struct SI {
+  static constexpr double Length = 1.;      // meter
+  static constexpr double Mass = 1.;        // kilogram
+  static constexpr double Time = 1.;        // second
+  static constexpr double Temperature = 1.; // Kelvin
+  static constexpr double Current = 1.;     // Amp
+  static constexpr double Charge = 1.;      // Coulomb
+  static constexpr double Capacitance = 1.; // Farad
+  static constexpr double Angle = 1.;       // Radian
+};
+
+struct CGS {
+  static constexpr double Length = 1.e2;                    // centimeter
+  static constexpr double Mass = 1.e3;                      // gram
+  static constexpr double Time = 1.;                        // second
+  static constexpr double Temperature = 1.;                 // Kelvin
+  static constexpr double Current = 1.e-1;                  // Biot
+  static constexpr double Charge = 2.997924580e9;           // Statcoulomb
+  static constexpr double Capacitance = 8.9831483395497e11; // Statfarad
+  static constexpr double Angle = 1.;                       // Radian
+};
+
+// Defines and encapsulates physical and mathematical constants in a purely
+// constexpr way, with support for multiple unit systems.
+template <typename UNITSYSTEM>
+class PhysicalConstants {
+ private:
```

Comment:
I think these should be `protected` so that `PhysicalConstants can be extended. For my application I would need to add some constants that are only applicable for astrophysics.

Reply:
That's a good point I hadn't appreciated, thanks for bringing it up -- I've made this change.